File: contest leetcode/weekly 190/Pseudo-Palindromic Paths in a Binary Tree.py
import collections
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: TreeNode) -> int:
        def palinChecker(text):
            hashMap = collections.Counter(text)
            check = False
            checkPalin = True
            for num in hashMap:
                if hashMap[num]&1 and not check:
                    check = True
                elif not hashMap[num] & 1:
                    continue
                else:
                    checkPalin = False
                    break
            return checkPalin

        def getPseudoPalinCnt(node, string):
            if not node:
                return
            string += str(node.val)
            if not node.left and not node.right:
                if palinChecker(string):
                    return 1
                else:
                    return 0
            leftChildCnt = getPseudoPalinCnt(node.left, string)
            rightChildCnt = getPseudoPalinCnt(node.right, string)
            return leftChildCnt + rightChildCnt
        return getPseudoPalinCnt(root, "")

File: December 30 day challenge/PseudoPalindromicPathsInABinaryTree.java
import java.io.*;
import java.util.*;

public class PseudoPalindromicPathsInABinaryTree {
    int [] freq;
    int cnt = 0;
    public void dfs(TreeNode node){
        if(node == null)
            return;
        freq[node.val]++;
        if(node.left == null && node.right == null){
            int totalLength = 0, oddCnt = 0;
            for(int num : freq){
                totalLength += num;
                if(num % 2 == 1)
                    oddCnt++;
            }
            if(totalLength % 2 == 0 && oddCnt == 0)
                cnt++;
            else if(totalLength % 2 == 1 && oddCnt == 1)
                cnt++;
        }
        if(node.left != null)
            dfs(node.left);
        if(node.right != null)
            dfs(node.right);
        freq[node.val]--;
    }
    public int pseudoPalindromicPaths (TreeNode root) {
        freq = new int[10];
        dfs(root);
        return cnt;
    }
}
